### DISCUSSION TITLE
Compatibility between devices (Apple and Microsoft have incompatible encodings)

### DISCUSSION THREAD

**Discussion by hakavlad** [samuel-lucas6/Kryptor#78] [General]
>All password inputs must be converted into Unicode Normalization Form KC, NFKC, and then encoded as UTF-8 bytes. This is to ensure compatibility between devices (Apple and Microsoft have incompatible encodings).

https://github.com/covert-encryption/covert/blob/main/docs/Specification.md#argon2-password-hashing

What do you think about it? Why doesn't Kryptor do Normalization?

**Comment by samuel-lucas6** [comment]
It sounds like that's a sensible idea (it's [recommended by NIST](https://pages.nist.gov/800-63-3/sp800-63b.html)), although it's unclear to me how big of an issue this is. It might be problematic with non-English keyboards, which is something I've never tested.

That just never occurred to me. It doesn't seem that widely discussed, and it looks like age and probably lots of programs don't do normalization either.

**Comment by hakavlad** [comment]
See also:

https://datatracker.ietf.org/doc/html/rfc8265
Preparation, Enforcement, and Comparison of Internationalized Strings Representing Usernames and Passwords
RFC 8265
RFC - Proposed Standard 
"It represents the consensus of the IETF community."

```
   4.  Normalization Rule: Unicode Normalization Form C (NFC) MUST be
       applied to all strings.
```

>It doesn't seem that widely discussed

However, it has been discussed for over 20 years (at least since RFC 3454 and RFC 4013) and is widely recommended by NIST and IETF.

**Reply by samuel-lucas6** [reply]
> However, it has been discussed for over 20 years (at least since RFC 4013) and is widely recommended by NIST and IETF.

Regrettably, I'm not omniscient. There are thousands of standards and documents, each tens or hundreds of pages long. It's completely unrealistic to be aware of or care about all of them.

I've heard of the NIST standard, so I should've at least skim read that (there's lots of irrelevant information). However, I've never heard of those RFCs.

People at NIST [didn't](https://github.com/usnistgov/800-63-3/issues/301) originally include that in NIST SP 800-63B, it's not in NIST SP 800-132, it's not mentioned on the OWASP password hashing page, it's not mentioned in the Argon2/scrypt/PBKDF2 RFCs, `string.Normalize()` is barely mentioned in the dotnet/runtime repo, there's no mention of normalization on the `Encoding.GetBytes()` documentation, etc. It could clearly be advertised better. It seems to mostly come up  if you search 'normalization', which you wouldn't do if you weren't aware of this problem in the first place.

**Reply by hakavlad** [reply]
>I'm not omniscient.

OK. Let's just stick to best practices going forward.

**Reply by samuel-lucas6** [reply]
I'm not convinced normalization is actually the solution here because it doesn't completely solve the problem. The better approach is probably to restrict the input to ASCII characters, which is I believe what a lot of websites do. Something to think about for the next major version.

**Reply by hakavlad** [reply]
>restrict the input to ASCII characters, which is I believe what a lot of websites do

It's just bad practice. Restricting character selection when entering a password:

- makes the selection inconvenient for those who prefer to use any characters;
- reduces the entropy of the password in any case (the adversary can avoid brute-forcing non-ASCII passwords if they are not allowed).

>it doesn't completely solve the problem

Normalization solves exactly the problem it is intended to solve.

>age and probably lots of programs don't do normalization

Magic Wormhole [uses NFC normalization](https://github.com/magic-wormhole/magic-wormhole/blob/0.17.0/src/wormhole/util.py#L33). This is an example of an application that follows best practices in this matter.

See also:

>Allow usage of all characters including unicode and whitespace. 

OWASP: [Authentication Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html#implement-proper-password-strength-controls)

>All printing ASCII [[RFC 20]](https://pages.nist.gov/800-63-3/sp800-63b.html#RFC20) characters as well as the space character SHOULD be acceptable in memorized secrets. Unicode [[ISO/ISC 10646]](https://pages.nist.gov/800-63-3/sp800-63b.html#ISOIEC10646) characters SHOULD be accepted as well.

NIST: [NIST SP 800-63B](https://pages.nist.gov/800-63-3/sp800-63b.html#memsecret)

**Reply by hakavlad** [reply]
>it's not mentioned on the OWASP password hashing page

This is mentioned elsewhere:

>The primary means of input validation for free-form text input should be:
>    - **Normalization:** Ensure canonical encoding is used across all the text and no invalid characters are present.

[Secrets Management Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#validating-free-form-unicode-text)

**Reply by hakavlad** [reply]
Also:

>Input validation
>  - Specify character sets, such as UTF-8, for all input sources (**canonicalization**)

https://owasp.org/www-project-secure-coding-practices-quick-reference-guide/stable-en/02-checklist/05-checklist#input-validation

>**Canonicalize:** To reduce various encodings and representations of data to a single simple form.

https://owasp.org/www-project-secure-coding-practices-quick-reference-guide/stable-en/03-appendices/05-glossary

**Reply by samuel-lucas6** [reply]
You make some good points, but a) most people won't use non-ASCII characters (e.g., randomly generated passwords/passphrases), b) the entropy argument doesn't matter if you choose a sufficiently strong password, c) normalization doesn't completely eliminate this problem as I've already said (it only improves the situation rather than completely fixing it), and d) the standards/guidance only ever say SHOULD rather than MUST. It may be best practice, but it seems like quite a bad solution if it's not properly fixing the problem. It's also annoying in the sense that you're creating more copies of the string password in memory.

**Reply by hakavlad** [reply]
>it's not properly fixing the problem

Your "complete" solution is simply to cut off a sore head.

>*most* people won't use non-ASCII characters

The DEI department condemns your decision! Think of minorities!

**Reply by hakavlad** [reply]
https://github.com/FiloSottile/age/discussions/601 Let's expand the number of participants in this discussion.

**Reply by samuel-lucas6** [reply]
> Your "complete" solution is simply to cut off a sore head.

Blame Unicode. You either ignore this problem, do a partial fix, or do a complete fix. And different people/organisations recommend different approaches to doing the partial fix so it's not even like there's one agreed method, whereas there's a clear complete fix.

> The DEI department condemns your decision! Think of minorities!

A lot of people speak/write English, ASCII characters are probably accessible on every device, some websites are more restrictive than ASCII characters, and people should be using randomly generated passwords/passphrases anyway, which use ASCII characters unless you go out of your way to use non-ASCII (e.g., a non-English BIP-39 wordlist). The EFF/Diceware/standard BIP-39 wordlists are in English.

Maybe I'll change my mind, but this problem doesn't have a pretty solution.

**Reply by hakavlad** [reply]
>the standards/guidance only ever say SHOULD rather than MUST

rfc8265 says MUST

**Reply by hakavlad** [reply]
>The better approach is probably to restrict the input to ASCII characters, which is I believe what a lot of websites do. Something to think about for the next major version.

How are you going to do this? Reject with an error phrases containing non-ASCII? I would consider such behavior of the tool to be offensive and discriminatory, unreasonably limiting my freedom to choose a password.

**Reply by samuel-lucas6** [reply]
> rfc8265 says MUST

Ok that's my bad; I misremembered. However, that RFC says something different to NIST and NIST only says it's optional, so who do you listen to?

> How are you going to do this? Reject with an error phrases containing non-ASCII? I would consider such behavior of the tool to be offensive and discriminatory, unreasonably limiting my freedom to choose a password.

Yes, that's how it would be done, although this issue is also problematic for file names, which probably shouldn't only be ASCII.

It's not discriminatory given this is a problem with Unicode that it sounds like literally cannot be fixed. You can use normalization and still have users fail to log in to a service/decrypt a file, which in my mind isn't a fix at all but a band-aid.

**Reply by hakavlad** [reply]
@samuel-lucas6 

> I'm not convinced normalization is actually the solution here because it doesn't completely solve the problem. The better approach is probably to restrict the input to ASCII characters

> c) normalization doesn't completely eliminate this problem as I've already said (it only improves the situation rather than completely fixing it)

> this problem doesn't have a pretty solution.

> this is a problem with Unicode that it sounds like literally cannot be fixed. You can use normalization and still have users fail to log in to a service/decrypt a file, which in my mind isn't a fix at all but a band-aid.


Look closer at RFC 8265:

>   All code points and blocks not explicitly allowed in the PRECIS
>   FreeformClass are disallowed; this includes private-use code points,
>   surrogate code points, and the other code points and blocks defined
>   as *"Prohibited Output"* in Section 2.3 of [RFC4013] (when corrected
>   per [Err1812]).

-- https://www.rfc-editor.org/rfc/rfc8265.html#section-4.1


```
2.3.  Prohibited Output

   This profile specifies the following characters as prohibited input:

      - Non-ASCII space characters [StringPrep, C.1.2]
      - ASCII control characters [StringPrep, C.2.1]
      - Non-ASCII control characters [StringPrep, C.2.2]
      - Private Use characters [StringPrep, C.3]
      - Non-character code points [StringPrep, C.4]
      - Surrogate code points [StringPrep, C.5]
      - Inappropriate for plain text characters [StringPrep, C.6]
      - Inappropriate for canonical representation characters
        [StringPrep, C.7]
      - Change display properties or deprecated characters
        [StringPrep, C.8]
      - Tagging characters [StringPrep, C.9]
```

-- https://www.rfc-editor.org/rfc/rfc4013#section-2.3

We can use Unicode safely if Prohibited Output is disallowed.

Here is an example of a library for Python: https://github.com/byllyfish/precis_i18n

**Reply by samuel-lucas6** [reply]
Thanks for the references. I'll take a look when I next implement something that involves the user providing a password. This problem isn't going to be fixed in Kryptor anytime soon as it's a breaking change, which means it'll be paired with other breaking changes.

**Comment by cipriancraciun** [comment]
@hakavlad thanks for pointing this out on the Age project (which I silently follow).

Following @samuel-lucas6's comments, I see that his main counterargument is the unreliability of non-ASCII passwords, which is in fact @hakavlad main purpose for Unicode normalization, and unfortunately on this subject (namely "passwords") I think I side with Samuel's argument.

For example, I've developed a [password generator tool](<https://github.com/volution/z-tokens>) (and a few related topics), and a few weeks ago I was pondering of adding Unicode-exclusive password patterns (i.e. just emojis or non ASCII-based passwords).  But I've decided against that because I think they'll generate more issues that they'll solve (for once due to possible poor support for Unicode inputs, and second for the portability issue).  (The main feature would have been increased entropy for the same "visual length", but as Samuel said, entropy can easily be increased with ASCII-only passwords.)

----

However, and the reason I've opened [my own ticket](<https://github.com/volution/z-tokens/issues/38>) on the subject, is that sometimes there might be cases where Unicode inputs (that do contribute cryptographically) could be encountered.

For example in the case of Associated Data (as part of AEAD) encryption schemes.  (Similar to Kryptor and Covert, I too have implemented as part of `z-tokens` an Age-like encryption tool.)

Now, I don't think this is the case for Kryptor (based on what I've read from the specification), but imagine that one adds a CLI argument like `--additional-data [some-human-readable-text]` where the user could specify the intended usage of the file (like for example, is this the `shutdown.bash` or the `startup.bash` script?), and would like this to be part of the authentication.  (This could protect against an attacker that swaps the two files.)

In such a case, the user is more likely to texts native to his own tongue, and thus Unicode.  Thus perhaps, especially in such situations Unicode normalization would be a requirement.

Getting back to Kryptor, it allows the usage of both a password and a secret key file (shared secret);  thus a user might just use the secret key file as the intended password, and the "password" as the "additional data" in a scheme similar to the above.

----

In the end, each developer makes a decision based on his priorities (which might be completely different from another's).  However I find this discussion quite interesting, at least from a theoretical point of view.

Perhaps, if a tool doesn't support Unicode normalization for a particular input, and that normalization is essential for portability (or cryptography), perhaps that tool should issue a warning or error on the usage of non ASCII inputs?

----

How about CLI tools (in Linux/BSD environments) that receive C-strings as inputs (i.e. byte arrays terminated by `\0`) via `main` (either as arguments or as environment variables);  how should these tools treat these inputs?  As UTF-8 encoded strings?  As byte encoded Unicode strings according to `LC_*` environment variables?

**Reply by hakavlad** [reply]
I see the benefit of Unicode in that people whose native language is not English can more easily memorize phrases in their own language. It is even more difficult for those who do not speak English well to memorize phrases in that language. I believe there are many such people in the world.

Some alphabets do not even require normalization and are not affected by this problem. For example, the Russian alphabet does not require normalization; all 33 letters have a standard representation in UTF-8 as 2 bytes.

Thus, I agree with the recommendations of NIST and OWASP that the input of any printable characters should be allowed.

**Reply by cipriancraciun** [reply]
> I see the benefit of Unicode in that people whose native language is not English can more easily memorize phrases in their own language.

I don't want to argue on the topic of internationalization and non-English centrist aspect of user interaction.

However, I just want to highlight that many decisions have to be balanced against risks and benefits.  The benefits are clear in the case of internationalization;  however has anyone made a thorough analysis of the risks involved?

----

From my own personal experience with IT support, I had a situation when one of our users suspected us of breaking into his account (via password resets), because each time when he went home he had to reset his password, and then when coming back to the office to reset it again, as each time the password he has chosen didn't work.

After many back-and-forth discussions, I asked the user if perhaps he was using any non English letters in his password, which it seems he did...

And this was with Windows AD-based login, thus not some random open-source application...

----

Back to the topic of allowing non-ASCII characters in a password, and why I view this as a potential risk for the user:
* imagine a non very tech-savvy user uses one OS to encrypt some data, and decides to use a special diacritic letter native to his tongue;
* now imagine he switches the OS (or even the version) to a different one, and tries do decrypt his data, and it gets a "password mismatch error" (because the MAC doesn't match);
* he types the password in Notepad and **visually** checks its exactly as he remembers;  (and we assume he didn't forget it;)
* what does he do now?  loses all its data?

When would something like this happen?  Let's take my native language, Romanian, for which unfortunately even the same OS, at different releases, for the same particular keystroke yields different Unicode code-points; see this Wikipedia section about the topic:
* <https://en.wikipedia.org/wiki/Romanian_alphabet#Comma-below_(%C8%99_and_%C8%9B)_versus_cedilla_(%C5%9F_and_%C5%A3)>

**Reply by hakavlad** [reply]
>however has anyone made a thorough analysis of the risks involved?

Maybe NIST and OWASP did. They probably give recommendations for a reason.

**Reply by hakavlad** [reply]
>now imagine he switches the OS (or even the version) to a different one, and tries do decrypt his data, and it gets a "password mismatch error" (because the MAC doesn't match)

Normalization solves exactly this problem.

**Reply by hakavlad** [reply]
>When would something like this happen? Let's take my native language, Romanian, for which unfortunately even the same OS, at different releases, for the same particular keystroke yields different Unicode code-points

Have you tried to solve the problem by normalization? Have there been cases where normalization did not solve the problem?

**Reply by cipriancraciun** [reply]
> Normalization solves exactly this problem.
> [...]
> Have you tried to solve the problem by normalization? Have there been cases where normalization did not solve the problem?

No, Unicode normalization does not solve this particular problem, exactly because it's not a Unicode normalization problem, where the same code-point (or sequence of code-points) can be encoded by multiple different sequences of bytes.

In the cited example, namely that of some Romanian diacritics, former Windows (and Linux) versions used to use a Turkish character that was **visually** similar to the actual diacritic that should have been used.  (Only later versions corrected the keymaps.)

Thus the actual issue was in a completely different part of the software stack, the input layer, and the actual software taking the input had no way to know that the Turkish letter was actually meant as a Romanian letter...

**Reply by hakavlad** [reply]
I see a solution in warning users that the use of certain Unicode characters may cause issues. I do not think it is right to restrict the freedom of some because certain input options may cause problems for others.

**Reply by hakavlad** [reply]
>later versions corrected the keymaps

So the problem is solved! It's better to ban old OSes, not Unicode.

**Reply by samuel-lucas6** [reply]
Thanks for joining the discussion @cipriancraciun. A warning in the documentation is definitely required. I assume you both also mean a warning in terms of checking for Unicode characters in the password, which is certainly worth investigating but may be complicated by the interactive password entry.

@hakavlad I really do understand where you're coming from; however, if the choice is between freedom and correctness, correctness wins. I don't think Unicode and its complexity can be defended. If the standard was better, countless bugs related to this issue would've been avoided.

If you apply normalization, this hasn't really been fixed because normalization doesn't correct for every character. Therefore, users can potentially shoot themselves in the foot, especially given most people have no awareness of this problem and it's not obvious what normalization doesn't fix. Furthermore, you end up giving a warning to be careful using Unicode characters, which discourages people from using Unicode anyway.

As for non-English speakers, as I've already said, there are much stricter password requirements on many websites, meaning they should already be used to using ASCII characters for passwords. Additionally, you should almost never be memorising passwords; you should be using a password manager, memorising the password for that, and randomly generating passwords for websites/applications using the password manager. If you do this using something like 1Password or Bitwarden, you'll get ASCII passwords. The only other time I'd recommend memorising a password is for disk encryption so you can access your device. When you're using a file encryption tool, you'll probably have access to your password manager so memorisation is unnecessary whilst worsening security. And memorised passwords should also be randomly generated for them to be strong, but there are obvious length/complexity limits due to memorisation.

**Reply by samuel-lucas6** [reply]
Another important point I've just realised is that websites that do support Unicode passwords have recovery options, whereas an offline application doesn't. Thus, it's more risky supporting Unicode passwords in an offline application. These standards/guidelines may be more geared towards online services.

**Reply by hakavlad** [reply]
>if the choice is between freedom and correctness

Granting freedom does not necessarily violate correctness; there is no contradiction between them.

With freedom, users can take care of correctness themselves. I believe that a combination of normalization and user warning is sufficient.

**Reply by hakavlad** [reply]
>giving a warning to be careful using Unicode characters, which discourages people from using Unicode anyway.

discourages is better then enforcing

**Reply by hakavlad** [reply]
>Additionally, you should almost never be memorising passwords; you should be using a password manager

"My house burned down, my devices burned down, I couldn't access my cloud storage and my password manager..."
https://shkspr.mobi/blog/2022/06/ive-locked-myself-out-of-my-digital-life 
This story also shows that remembering passwords can be useful.

In fact, everywhere I say "passwords", I mean "passphrases".

**Reply by hakavlad** [reply]
>you'll probably have access to your password manager

I want to be able to download backups from a public storage and successfully decrypt them after I lose all my devices.

**Reply by samuel-lucas6** [reply]
> With freedom, users can take care of correctness themselves.

That's exactly the problem. The onus shouldn't be on the user. Something like logging in to a site/being able to decrypt should always work (assuming everything else is correct/not tampered with). Your argument feels like 'let's not do validation because the user can do validation themselves', which is generally poor design.

> discourages is better then enforcing

Not if you want to guarantee that this isn't a problem. In that case, it's objectively worse.

However, to be honest, I can get onboard with that because that's what lots of other people seem to be doing without caring. If that's 'best practice' and keeps people happy, it's arguably good enough.

The biggest thing that bothers me about normalization is that it's duplicating the password in memory. Now the password already gets duplicated/not cleared properly in Kryptor, but I tried to ensure the password was erased in [Cahir](https://github.com/samuel-lucas6/Cahir). With any sort of non-string password entry, you have to convert it to a string for normalization, with the [normalization function](https://learn.microsoft.com/en-us/dotnet/api/system.string.normalize?view=net-8.0) also returning a new string. Strings are immutable and not meant to be zeroed as it can potentially crash the runtime, although it seemed to work ok in Cahir. But the point is that you want to minimise the duplication of secrets to limit the risk of them being retrieved/dumped. Normalization defeats half the point of doing an interactive password entry.

> This story also shows that remembering passwords can be useful.

I'm not disputing that, but you can't remember passwords for everything. If you do, they'll be terrible. It makes more sense to back up access to your password manager. For example, if you're concerned about MFA, you keep a backup of the recovery code in a safe deposit box at a bank, a box buried in the garden, hidden somewhere at a family member/friend's house, etc. There are ways it can be done, just they're not easy.

> I want to be able to download backups from a public storage and successfully decrypt them after I lose all my devices.

Except unless you remember your cloud storage account password and recovery code, you won't have access to that anyway. And it's virtually impossible to remember recovery codes because a) you never use them, b) they're typically gibberish, and c) they're typically long.

But this discussion is going off topic.

**Reply by cipriancraciun** [reply]
> The biggest thing that bothers me about normalization is that it's duplicating the password in memory.

Indeed, the major concern with regard to cryptographic material (like passwords) is erasing their traces after usage, in order to avoid exfiltration.  And this is a topic I've thought at length about how to eliminate copies of a password from memory.

And, unfortunately, the conclusion I've reached is that it's almost impossible to make sure copies of the password aren't left anywhere on the system, because:
* if the tool is a CLI, then the password has already traversed the keyboard buffer (in the kernel), X or Wayland (or the user-interface system), GTK or QT (or the toolkit being used), the TTY buffer (in the kernel for the TTY emulation), and finally the CLI process;
* if the tool is a GUI, then it's almost as above, less a few pieces from the stack;
* if the password is supplied from another process, say via pipes, it's still present in the kernel's memory;

Besides zero-ing the buffers in our own systems, we could do the following:
* use allocators that zero on free for all involved software;  (this adds performance penalties;)
* use kernel flags (at least on Linux) that zero all memory (used in buffers, pipes, etc.) after use;  (which is not enabled by default due to performance penalties;)
* disable SWAP;  (one could try to pin a processes memory, but by default non-root users can only do so for a couple of KB;)

Thus, at this point, with today's software, I don't think duplicate copies of a string in a process's memory is much of a concern.

**Reply by hakavlad** [reply]
>The onus shouldn't be on the user. 

It is impossible to absolve the user of responsibility.

Either way, the user is responsible for:
- the security of the environment in which the application is run;
- the strength of the keys used;
- preventing the leakage of keys and plaintexts.

>you can't remember passwords for everything

Of course, passwords for websites are stored in a password manager. But having the ability to conveniently memorize passphrases for encrypted files is just what we need.

>unless you remember your cloud storage account password and recovery code, you won't have access to that anyway

I was talking about public storage. It can be, for example, a *public* telegram channel (you can store files up to 2 GB free). If you lose your device, you won't need keys to access these files. So, this is exactly the case when it is important to be able to conveniently remember passphrases for encrypted files. This is a way to solve the problem described in the article.

>disable SWAP

On Linux, you can use ZRAM to create a compressed swap space, which helps prevent memory page swapping to disk. On Fedora Linux, this is [offered as a default option](https://fedoraproject.org/wiki/Changes/SwapOnZRAM).

**Comment by samuel-lucas6** [comment]
I've updated the website, adding a section in [Known limitations](https://www.kryptor.co.uk/security-limitations#unicode-passwords) and a warning on the Tutorial pages talking about using a passphrase (e.g., [this](https://www.kryptor.co.uk/tutorial/entering-a-passphrase) page).

I'll look into adding a warning in the program based on char values exceeding ASCII.

**Reply by hakavlad** [reply]
BTW w3c recommends NFC https://www.w3.org/TR/charmod-norm/ 
Also: 
>Specifications MUST allow a Unicode character encoding.

**Reply by hakavlad** [reply]
>To eliminate the problem, you have to enforce ASCII characters

It's a redundant solution, a redundant limitation.

It is sufficient not to use certain Unicode symbols when using the tool on different devices.

For example, I can safely use characters from the Russian alphabet; there are no problems with them, as well as with ASCII.

Additionally, I can use ANY Unicode characters as long as I do so on the same device and in the same manner.

Thus, you offer a simple but overly restrictive solution.

**Reply by samuel-lucas6** [reply]
This is getting a bit silly as we're going around in circles and having multiple separate conversations in different places.

> It's a redundant solution, a redundant limitation.

Whether it's redundant depends on the user.

> It is sufficient not to use certain Unicode symbols when using the tool on different devices.

Except you have to trust the user, who shouldn't have to know anything about Unicode, to know which Unicode characters to avoid. The average person, even the average computer science student, doesn't know enough about Unicode to do this.

> Additionally, I can use ANY Unicode characters as long as I do so on the same device and in the same manner.

I'm not sure that's correct if it depends on the keyboard, operating system, settings, etc. It sounds possible for you to update/upgrade your OS, change your keyboard/settings, whatever and this could potentially be a problem.

> Thus, you offer a simple but overly restrictive solution.

However, as I keep saying, it's the only actual solution. Normalization reduces but doesn't eliminate the risk. They're not the same thing.
